testing++: formatting, remove Prove, add sections (random checking) (#413)
LispCookbook/cl-cookbook: testing.md

integration services like GitHub Actions, Gitlab CI, Travis CI or Coveralls.

We will be using a mature testing framework called
[FiveAM](https://github.com/lispci/fiveam). It supports test suites,
random testing, test fixtures (to a certain extent) and, of course,
interactive development.

Previously on the Cookbook, the recipe was cooked with [Prove](https://github.com/fukamachi/prove). It used to be a widely liked testing framework but, because of some shortcomings, its repository was later archived. Its successor [Rove](https://github.com/fukamachi/rove) is not stable enough and lacks some features, so we didn't pick it. There are also some [other testing frameworks](https://github.com/CodyReichert/awesome-cl#unit-testing) to explore if you feel like it.

FiveAM has an [API documentation](https://common-lisp.net/project/fiveam/docs/index.html). You may inspect it or simply read the docstrings in code. Most of the time, they would provide sufficient information that answers your questions… if you didn't find them here. Let's get started.

## Testing with FiveAM

FiveAM has 3 levels of abstraction: check, test and suite. As you may have guessed:

> 1. A **check** is a single assertion that asserts a boolean value. The most used check is `is`, followed by an assertion of yours. For example, `(is (= 2 (+ 1 1)))`.

Reviewer: Too many times "assertion". Is it really a **check** in FiveAM? Or can we call it simply an assertion?

```suggestion
1. A **check** or assertion. The most used check is `is`, which asserts that its single argument is truthy. For example, `(is (= 2 (+ 1 1)))`.
```

Author: yes it is called a check in fiveam.

Author: "1. A **check** is a single assertion that checks that its argument is truthy. The most used check is `is`. For example, `(is (= 2 (+ 1 1)))`."

Reviewer: Looks great ;)

    (t nil)))
~~~

> We will test that code:
> 
> - we must ensure that the content read in a file is the expected content,
> - we must test that the condition is signaled if the file doesn't exist.

Reviewer: Sorry I don't understand this paragraph. This is what we'll test below?

If so:
```suggestion
Next, our test code shall:
- ensure that the content read in a file is the expected content,
- test that the condition is signaled if the file doesn't exist.
```

still, I'm a bit lost

Author: oh. mmh. "We will write tests for that code" (the code above). "In particular, we must ensure: that the content… - that a condition…".

Reviewer: Much clearer :+1:  :)

Author: great, thanks :)

The code below defines a suite named `my-system`. We will use it as the root suite for the whole system.

~~~lisp
(def-suite my-system
  :description "Test my system")
~~~

Then let's define another suite for testing the `read-file-as-string` function.

~~~lisp
;; Define a suite and set it as the default for the following tests.
(def-suite read-file-as-string
  :description "Test the read-file-as-string function."
  :in my-system)
(in-suite read-file-as-string)

;; Alternatively, the following line is a combination of the 2 lines above.
(def-suite* read-file-as-string :in my-system)
~~~

Here a new suite named `read-file-as-string` has been defined. It is declared to be a child suite of `my-system` as specified by the `:in` keyword. The macro `in-suite` sets it as the default suite for the tests defined later.

### Defining tests

Before diving into tests, here is a brief introduction of the available checks you may use inside tests:

> * The macro `is` might be the mostly used check. It simply checks if the given expression returns a true value and generates a `test-passed` or `test-failure` result accordingly.

Reviewer: "The macro `is` might be the most used check"
